### PR TITLE
Added some code to deal with hidden sheets

### PR DESCRIPTION
**Pull request by shachaf-ashkenazi**
Apparently those exists, and without this code if a hidden sheet exists before designated sheet, it wouldn't parse it.

### CODE REVIEW

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java [hunk 1]
```diff
@@ -86,15 +89,24 @@ private boolean isLastLine(boolean hasNext) {
...
+    private Optional<Iterator<Row>> getRowIterator(InputStream inputStream) {
         final Workbook workbook = StreamingReader.builder()
                 .rowCacheSize(10)
                 .bufferSize(1024)
                 .open(inputStream);
-        final Sheet sheet = workbook.getSheetAt(0);
-        final Iterator<Row> rowIterator = sheet.rowIterator();
-        
-        return rowIterator;
+        final Optional<Sheet> sheet = getFirstVisibleSheet(workbook);
```
Comment (mosheeshel) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: A definite improvement over using 0 magic number.

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java [hunk 2]
```diff
@@ -86,15 +89,24 @@ private boolean isLastLine(boolean hasNext) {
...
+        return sheet.map(Sheet::rowIterator);
+    }
+
+    private Optional<Sheet> getFirstVisibleSheet(Workbook workbook) {
+        OptionalInt firstIndex = IntStream.range(0, workbook.getNumberOfSheets())
+                .filter(i -> !isSheetHidden(i, workbook))
+                .findFirst();
+        return firstIndex.isPresent() ? Optional.of(workbook.getSheetAt(firstIndex.getAsInt())) : Optional.empty();
```
Comment (mosheeshel) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: Not a Java 8 expert but seems to me you can use flatMap to reduce some ifs

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java [hunk 3]
```diff
@@ -86,15 +89,24 @@ private boolean isLastLine(boolean hasNext) {
         return !hasNext;
     }
 
-    private Iterator<Row> getRowIterator(InputStream inputStream) {
+    private Optional<Iterator<Row>> getRowIterator(InputStream inputStream) {
         final Workbook workbook = StreamingReader.builder()
                 .rowCacheSize(10)
                 .bufferSize(1024)
```
Comment (mosheeshel) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: Why did we choose these values? Should they be controlled by configuration? Optional parameters in the API call?

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java [hunk 4]
```diff
@@ -86,15 +89,24 @@ private boolean isLastLine(boolean hasNext) {
...
                 .open(inputStream);
-        final Sheet sheet = workbook.getSheetAt(0);
-        final Iterator<Row> rowIterator = sheet.rowIterator();
-        
-        return rowIterator;
+        final Optional<Sheet> sheet = getFirstVisibleSheet(workbook);
+        return sheet.map(Sheet::rowIterator);
```
Comment (mosheeshel) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: Can inline
`return sheet.map(getFirstVisibleSheet(workbook)::rowIterator);`

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java [hunk 5]
```diff
@@ -55,10 +57,11 @@ public XlsxToCsvConverter(String delimiter) {
      * @throws IOException if an I/O error occurs.
      */
     public void convert(InputStream inputStream, FileOutputStream fileOutputStream) throws IOException {
-        final Iterator<Row> rowIterator = getRowIterator(inputStream);
-        if (!rowIterator.hasNext()) {
+        final Optional<Iterator<Row>> optionalRowIterator = getRowIterator(inputStream);
+        if (!optionalRowIterator.isPresent() || !optionalRowIterator.get().hasNext()) {
```
Comment (mosheeshel) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: Can replace most of this with
`rowIterator.orElseThrow(NoSheetFoundException)`
Can handle the empty sheet `!hasnext()` with another exception.
I think log is not enough (but using SLF4J is good)

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java (hunk 3 above)
Comment (shachaf-ashkenazi) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: It was a recommendation by the guy who wrote the streaming api. I think it just makes the user deal with things that are under the hood, but maybe we should supply a constructor with those parameters and a default one that just uses these ones. Sounds good?

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java (hunk 2 above)
Comment (shachaf-ashkenazi) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: It's an OptionalInt and it doesn't have flatMap

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java (hunk 3 above)
Comment (mosheeshel) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: an additional constructor (with a builder) sounds like a great option - we'll probably add more such options as we go ahead

Comment (shachaf-ashkenazi) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: Already added a builder to the pr

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java [hunk 6]
```diff
@@ -32,8 +31,11 @@
     private static final String COMMA = ",";
 
     private String delimiter = COMMA;
-    private CellHandler cellHandler = new CellHandler();
+    private int rowCacheSize = 10;
+
+    private int bufferSize = 1024;
 
+    private CellHandler cellHandler = new CellHandler();
```
Comment (mosheeshel) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: `final`?

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java [hunk 7]
```diff
@@ -50,20 +52,36 @@ public XlsxToCsvConverter(String delimiter) {
     }
 
     /**
+     *
+     * @param delimiter - a delimiter used to differentiate values in the translated file.
+     * @param rowCacheSize - cache size of rows to save in memory (for streaming purposes).
+     * @param bufferSize - buffer size to keep in memory (for streaming purposes).
+     */
+    public XlsxToCsvConverter(String delimiter, int rowCacheSize, int bufferSize) {
```
Comment (mosheeshel) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: this is why a builder pattern is better, you have "sane" defaults for everything and user only needs to change what he whats (without a special constructor for each case).

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java [hunk 8]
```diff
@@ -32,8 +31,11 @@
     private static final String COMMA = ",";
 
     private String delimiter = COMMA;
-    private CellHandler cellHandler = new CellHandler();
+    private int rowCacheSize = 10;
```
Comment (mosheeshel) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: change these to `private final `,  and initialize in default ctor from default values (constants).

File: src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java (hunk 2 above)
Comment (mosheeshel) [inline src/main/java/com/kenshoo/xlsx2csv/XlsxToCsvConverter.java]: You're right... :disappointed:

File: src/test/java/com/kenshoo/xlsx2csv/XlsxToCsvConverterTest.java [hunk 9]
```diff
@@ -28,16 +28,21 @@
...
         testFileTranslation("bulk_import_test_2_empty.xlsx", "bulk_import_expected_test_2_empty.csv");
     }
 
-    private void testFileTranslation(String fileNameToTranslate, String expectedFileName) {
+    @Test
+    public void testTranslateFileWithHiddenSheets() throws Exception {
+        testFileTranslation("bulk_import_test_3_with_hidden_sheets.xlsx", "bulk_import_expected_test_1.csv");
+    }
+
+    private void testFileTranslation(String fileNameToTranslate, String expectedFileName) throws Exception{
```
Comment (mosheeshel) [inline src/test/java/com/kenshoo/xlsx2csv/XlsxToCsvConverterTest.java]: rename function `translateFile`? it doesn't test, it just runs the scenario...

File: src/test/java/com/kenshoo/xlsx2csv/XlsxToCsvConverterTest.java [hunk 10]
```diff
@@ -28,16 +28,21 @@
...
 
     @Test
-    public void testTranslateFile() {
+    public void testTranslateFile() throws Exception {
         testFileTranslation("bulk_import_test_1.xlsx", "bulk_import_expected_test_1.csv");
     }
 
-    @Test
-    public void testTranslateEmptyFile() {
+    @Test(expected = IOException.class)
```
Comment (mosheeshel) [inline src/test/java/com/kenshoo/xlsx2csv/XlsxToCsvConverterTest.java]: I prefer using JUnit `ExpectedException` rule if you handle many exceptions
1. it allows compile time check of the test code
2. it supports more assertions and verification (for example you can assert on the exception message)

**Comment by shachaf-ashkenazi** [comment]
Addressed all comments

**Comment by shachaf-ashkenazi** [comment]
And added the builder to code usages in readme

**Pull request merged**